src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -26,27 +32,213 @@ const int32_t CompareOptionsIgnoreCase = 1;
+                if (hiraganaChar <= 0x3096 || hiraganaChar >= 0x309D) // characters between 3096 and 309D are not mapped to katakana
+                {
+                    customRules.push_back('&');
+                    customRules.push_back(hiraganaChar);
+                    customRules.push_back(compareChar);
+                    customRules.push_back(hiraganaChar + hiraganaToKatakanaOffset);
+                }
+            }
+        }
+
+        if (needsIgnoreWidthCustomRule || needsNotIgnoreWidthCustomRule)
+        {
+            UChar compareChar = needsIgnoreWidthCustomRule ? '=' : '<';
+            
+            UChar lowerChar;
+            for (int i = 0; i < g_HalfFullCharsLength; i++)
+            {
+                customRules.push_back('&');
+
+                // the lower chars need to be checked for escaping since they contain ASCII punctuation
+                lowerChar = g_HalfFullLowerChars[i];
+                if (NeedsEscape(lowerChar))
+                {
+                    customRules.push_back('\\');
+                }
+                customRules.push_back(lowerChar);
+
+                customRules.push_back(compareChar);
+                customRules.push_back(g_HalfFullHigherChars[i]);
+            }
+        }
+    }
+
+    return customRules;
+}
+
 /*
- * To collator returned by this function is owned by the callee and must be
+ * The collator returned by this function is owned by the callee and must be
  * closed when this method returns with a U_SUCCESS UErrorCode.
  *
  * On error, the return value is undefined.
  */
 UCollator* CloneCollatorWithOptions(const UCollator* pCollator, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pClonedCollator = ucol_safeClone(pCollator, nullptr, nullptr, pErr);
+    UColAttributeValue strength = UCOL_DEFAULT;
 
-    if ((options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase)
+    bool isIgnoreCase = (options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase;
+    bool isIgnoreNonSpace = (options & CompareOptionsIgnoreNonSpace) == CompareOptionsIgnoreNonSpace;
+    bool isIgnoreSymbols = (options & CompareOptionsIgnoreSymbols) == CompareOptionsIgnoreSymbols;
+
+    if (isIgnoreCase)
     {
-        ucol_setAttribute(pClonedCollator, UCOL_STRENGTH, UCOL_SECONDARY, pErr);
+        strength = UCOL_SECONDARY;
+    }
+
+    if (isIgnoreNonSpace)
+    {
+        strength = UCOL_PRIMARY;
+    }
+
+    UCollator* pClonedCollator;
+    std::vector<UChar> customRules = GetCustomRules(options, strength);
+    if (customRules.empty())
+    {
+        pClonedCollator = ucol_safeClone(pCollator, nullptr, nullptr, pErr);
+    }
+    else
+    {
+        int32_t customRuleLength = customRules.size();
+
+        int32_t localeRulesLength;
+        const UChar* localeRules = ucol_getRules(pCollator, &localeRulesLength);
+
+        std::vector<UChar> completeRules(localeRulesLength + customRuleLength + 1, '\0');
```

Comment:
The NULL terminator.  I was getting errors from ICU if this UChar* wasn't null terminated.